[Sema] Don't abort the result builder transform if the builder contained an ErrorExpr

ahoppen opened the pull request:
When matching a result builder in the constraint system, the closure has already been pre-checked so we don’t need to pre-check it again. Also, we can continue type checking even if the closure contained an `ErrorExpr` since result builders are now type checked like regular closures.

ahoppen commented:
@swift-ci Please smoke test

xedin commented on lib/Sema/BuilderTransform.cpp:
```diff
@@ -2462,27 +2462,23 @@ ConstraintSystem::matchResultBuilder(AnyFunctionRef fn, Type builderType,
-      return getTypeMatchFailure(locator);
-
-    return getTypeMatchSuccess();
+    assert(false &&
```
`llvm_unreachable` is better here.

ahoppen committed:
[Sema] Don't abort the result builder transform if the builder contained an ErrorExpr

When matching a result builder in the constraint system, the closure has already been pre-checked so we don’t need to pre-check it again. Also, we can continue type checking even if the closure contained an `ErrorExpr` since result builders are now type checked like regular closures.

ahoppen commented:
set

ahoppen commented:
@swift-ci Please smoke test

xedin reviewed (approved):
Thank you!

ahoppen commented:
@swift-ci Please SourceKit stress test
